### PR TITLE
RFC: Add final support for compress audio 

### PR DESCRIPTION
**Pull request by dbaluta**
This patch series adds the final support patches for compress audio playback. 
It mostly touches the codec_adapter() copy function which was initially designed only for PCM frames. 

This patches are sent for an early review. They will eventually get merged after I will sort out @cujomalainey PR: https://github.com/thesofproject/sof/pull/3881

@mrajwa @paulstelian97 @iuliana-prodan @cujomalainey @stolx please have a look.

@cujomalainey I will start working on sorting out https://github.com/thesofproject/sof/pull/3881 from now on.

### CODE REVIEW

File: src/audio/codec_adapter/codec_adapter.c [hunk 1]
```diff
@@ -245,7 +245,7 @@ static int codec_adapter_prepare(struct comp_dev *dev)
 	}
 
 	/* Allocate local buffer */
-	buff_size = MAX(cd->period_bytes, codec->cpd.in_buff_size) * buff_periods;
+	buff_size = MAX(cd->period_bytes, codec->cpd.out_buff_size) * buff_periods;
```
Comment (paulstelian97) [inline src/audio/codec_adapter/codec_adapter.c:248]: What type of data does the buffer actually hold? The input to the algorithm or the output from it? I hope this won't break when we eventually deal with the compress algorithms (compared to the decompress ones we're handling now).

File: tools/topology/sof/pipe-codec-adapter-playback.m4 [hunk 2]
```diff
@@ -115,10 +115,10 @@ W_CODEC_ADAPTER(0, PIPELINE_FORMAT, DAI_PERIODS, DAI_PERIODS, CA_SCHEDULE_CORE,
         LIST(`          ', "CA_SETUP_CONTROLBYTES_NAME_PIPE", "CA_RUNTIME_CONTROLBYTES_NAME_PIPE"))
 
 # Playback Buffers
-W_BUFFER(0, COMP_BUFFER_SIZE(DAI_PERIODS,
+W_BUFFER(0, COMP_BUFFER_SIZE(8,
```
Comment (paulstelian97) [inline tools/topology/sof/pipe-codec-adapter-playback.m4:118]: We need to figure out what's with this hardcoding.

File: tools/topology/sof-imx8-processing-wm8960.m4 [hunk 3]
```diff
@@ -0,0 +1,91 @@
+#
+# Topology for i.MX8MP board with wm8960 codec
```
Comment (paulstelian97) [inline tools/topology/sof-imx8-processing-wm8960.m4:2]: Why is the topology called "processing", not "codec_adapter" or anything related?

File: src/audio/codec_adapter/codec_adapter.c [hunk 4]
```diff
@@ -434,6 +434,9 @@ static int codec_adapter_copy(struct comp_dev *dev)
 		bytes_to_process -= codec->cpd.consumed;
 		processed += codec->cpd.consumed;
 		produced += codec->cpd.produced;
+
+		audio_stream_produce(&local_buff->stream, produced);
```
Comment (iuliana-prodan) [inline src/audio/codec_adapter/codec_adapter.c:438]: I don't follow your reasoning here.

If you update in each iteration the `processed` and `produced` and then update the pointers in `local_buff` and `source `buffer, wouldn't that mean that you process some bytes twice?

In the first iteration you decode x1 bytes and get y1 bytes = >` processed = x1` and` produced = y1`.
In next iteration you decode x2 and get y2 => `processed = x1+x2`   and `produced  = y1 + y2`.
That means that x1 and y1 are taken into account twice when updating the buffers pointers.

File: tools/topology/sof-imx8-processing-wm8960.m4 (hunk 3 above)
Comment (iuliana-prodan) [inline tools/topology/sof-imx8-processing-wm8960.m4:2]: If this topology is for 8QM and 8QXP maybe you should remove/update the i.MX8MP comment.

**Review by iuliana-prodan — CHANGES_REQUESTED** [review]

File: src/audio/codec_adapter/codec_adapter.c (hunk 1 above)
Comment (dbaluta) [reply src/audio/codec_adapter/codec_adapter.c:248]: This buffer is used to copy the algorithm output buffer. Because usually we cannot consume all of it in one period we need to store it locally.

File: tools/topology/sof/pipe-codec-adapter-playback.m4 (hunk 2 above)
Comment (dbaluta) [reply tools/topology/sof/pipe-codec-adapter-playback.m4:118]: Yes, yes! Will fix it asap.

File: tools/topology/sof-imx8-processing-wm8960.m4 (hunk 3 above)
Comment (dbaluta) [reply tools/topology/sof-imx8-processing-wm8960.m4:2]: @paulstelian97 I think I followed the initial naming from Intel. I could change it with a better naming. Codec_adapter is more of an internal naming users should care about this.

File: src/audio/codec_adapter/codec_adapter.c [hunk 5]
```diff
@@ -406,6 +406,13 @@ static int codec_adapter_copy(struct comp_dev *dev)
 			break;
 		}
 
+		/* TODO: remove magic number and use possible maximum output per
+		 * decoding algorithm type
+		 * 9216 is maximum frame length bytes for mp3 at 192Khz, 24bits,
+		 */
+		if (local_buff->stream.free < 9217)
```
[the remark this answers is not in the document]
Comment (dbaluta) [reply src/audio/codec_adapter/codec_adapter.c:413]: @thanks Iulia for the comment. What I actually plan to do is to create a generic function that given a codec and its parameters it will return the maximum amount of the output data.

Something like `code_adapter_get_max_bytes`? I just couldn't find a good name. This work is still RFC. We won't submit the code like that.

File: src/audio/codec_adapter/codec_adapter.c (hunk 4 above)
Comment (dbaluta) [reply src/audio/codec_adapter/codec_adapter.c:438]: Good catch. Actually what I meant is this:

```
-               audio_stream_produce(&local_buff->stream, produced);
-               comp_update_buffer_consume(source, processed);
+               audio_stream_produce(&local_buff->stream, codec->cpd.produced);
+               comp_update_buffer_consume(source,codec->cpd.consumed);
```

File: tools/topology/sof-imx8-processing-wm8960.m4 (hunk 3 above)
Comment (dbaluta) [reply tools/topology/sof-imx8-processing-wm8960.m4:2]: Yes, I think that's a good idea. Copy paste error :).

File: src/audio/codec_adapter/codec_adapter.c (hunk 1 above)
Comment (paulstelian97) [reply src/audio/codec_adapter/codec_adapter.c:248]: Then it would probably better to call it `local_output_buffer` or something.

File: tools/topology/sof-imx8-processing-wm8960.m4 (hunk 3 above)
Comment (paulstelian97) [reply tools/topology/sof-imx8-processing-wm8960.m4:2]: Maybe "compress" or something.

File: src/audio/codec_adapter/codec_adapter.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (iuliana-prodan) [reply src/audio/codec_adapter/codec_adapter.c:413]: A generic function sounds good. 
Regarding the naming, I think it's ok, but I'm not good with names either :)

[the remark this answers is not in the document]
Comment (lgirdwood) [reply src/audio/codec_adapter/codec_adapter.c:413]: Fwiw the 9212 should also be a topology 2.0 attribute (as this can also be enforced in any MP3 topology)

**Review by lgirdwood — COMMENTED** [review]
I assume we will support full passthrough mode for this work so this can be validated by CI without binary codecs ?

File: src/audio/codec_adapter/codec_adapter.c (hunk 1 above)
Comment (dbaluta) [reply src/audio/codec_adapter/codec_adapter.c:248]: @paulstelian97 I won't modify the existing buffer name. This is a good name choose by the initial commit of @mrajwa

File: src/audio/codec_adapter/codec_adapter.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (dbaluta) [reply src/audio/codec_adapter/codec_adapter.c:413]: @lgirdwood this could be an option. It looks like at least for mp3 the number of samples resulted after decoding one frame is always the same 1152 (9217 = 1152 samples * 2 channels * 4 bytes / sample + 1). If this holds true for all the decoding algorithms then getting this via topology could be an option.

**Comment by dbaluta** [comment]
@lgirdwood yes we will support pass through this is one of our hopes to test the codec_adapter publicly without private binary blobs.
We need to clarify some things about pass through:
* currently there is a simple pass through family of codecs [1]. This is already merged and I will take care that current patches do not break it.
* there is a proposal to add a pass through codec type inside Cadence family of codecs [2]

 @johnylin76  is there any update for [2].

[1] https://github.com/thesofproject/sof/blob/master/src/audio/codec_adapter/codec/passthrough.c 
[2] https://github.com/thesofproject/sof/pull/3766

File: src/audio/codec_adapter/codec_adapter.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (plbossart) [reply src/audio/codec_adapter/codec_adapter.c:413]: it's not true for the lower frequencies, the result is 576 samples.

More generically, there's actually a need to decouple input frames and output 'granules'. Sometimes frames are consumed without generating any data (e.g. to check for sync words or to build a history), and conversely on stream close there may still be data in a history buffer that needs to be flushed even with no incoming data. In general though the output granularity is constant and defined by the filter banks.

File: src/audio/codec_adapter/codec_adapter.c [hunk 6]
```diff
@@ -406,6 +406,13 @@ static int codec_adapter_copy(struct comp_dev *dev)
 			break;
 		}
 
+		/* TODO: remove magic number and use possible maximum output per
+		 * decoding algorithm type
+		 * 9216 is maximum frame length bytes for mp3 at 192Khz, 24bits,
```
Comment (plbossart) [inline src/audio/codec_adapter/codec_adapter.c:411]: that makes no sense. mp3 never supported 192kHz and perceptual coders don't care about bit depth.

I don't recall what the max value was but this was something at 320 kbits/s

File: tools/topology/sof-imx8-processing-wm8960.m4 [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+# playback DAI is SAI3 using 2 periods
+# Buffers use s32le format, with 48 frame per 1000us on core 0 with priority 0
+DAI_ADD(sof/pipe-dai-playback.m4,
+	1, SAI, 1, sai1-wm8960-hifi,
+	PIPELINE_SOURCE_1, 2, s32le,
+	1000, 0, 0, SCHEDULE_TIME_DOMAIN_DMA)
+
+
+# PCM Low Latency, id 0
```
Comment (plbossart) [inline tools/topology/sof-imx8-processing-wm8960.m4:80]: PCM?

File: tools/topology/sof-imx8-processing-wm8960.m4 [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+dnl PCM_DUPLEX_ADD(name, pcm_id, playback, capture)
+COMPR_PLAYBACK_ADD(Port0, 0, PIPELINE_PCM_1)
```
Comment (plbossart) [inline tools/topology/sof-imx8-processing-wm8960.m4:83]: CompressedPort0?

File: src/audio/codec_adapter/codec_adapter.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (dbaluta) [reply src/audio/codec_adapter/codec_adapter.c:413]: @plbossart 

image

File: tools/topology/sof-imx8-processing-wm8960.m4 (hunk 7 above)
Comment (dbaluta) [reply tools/topology/sof-imx8-processing-wm8960.m4:80]: @plbossart copy paste of course.

File: tools/topology/sof-imx8-processing-wm8960.m4 (hunk 8 above)
Comment (dbaluta) [reply tools/topology/sof-imx8-processing-wm8960.m4:83]: Makes sense.

File: src/audio/codec_adapter/codec_adapter.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (plbossart) [reply src/audio/codec_adapter/codec_adapter.c:413]: I think it would be a grave mistake to use a single period. 

if the scheduling is constrained by the output size, then it may happen than you have to decode multiple frames in the allocated time. In our past integration, the decoder was part of a different pipeline and scheduled with a lower priority, so that there were no peaks in the MCPS usage. I really think you have to decouple the codec_adapter from the dai scheduling....

File: src/audio/codec_adapter/codec_adapter.c (hunk 6 above)
Comment (dbaluta) [reply src/audio/codec_adapter/codec_adapter.c:411]: @plbossart It is clear that I still need to read upon mp3 stuff and all. I'm still confused about all the version / etc. 

I started with a sample mp3 file:

```
root@imx8qxpc0mek:~# file samples/lb.mp3    
samples/lb.mp3: MPEG ADTS, layer III, v1, 192 kbps, 48 kHz, Stereo
```

From the internet I have this:

```
// Bitrates - use [version][layer][bitrate]
const uint16_t mpeg_bitrates[4][4][16] = {
  { // Version 2.5
    { 0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0, 0 }, // Reserved
    { 0,   8,  16,  24,  32,  40,  48,  56,  64,  80,  96, 112, 128, 144, 160, 0 }, // Layer 3
    { 0,   8,  16,  24,  32,  40,  48,  56,  64,  80,  96, 112, 128, 144, 160, 0 }, // Layer 2
    { 0,  32,  48,  56,  64,  80,  96, 112, 128, 144, 160, 176, 192, 224, 256, 0 }  // Layer 1
  },
  { // Reserved
    { 0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0, 0 }, // Invalid
    { 0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0, 0 }, // Invalid
    { 0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0, 0 }, // Invalid
    { 0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0, 0 }  // Invalid
  },
  { // Version 2
    { 0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0, 0 }, // Reserved
    { 0,   8,  16,  24,  32,  40,  48,  56,  64,  80,  96, 112, 128, 144, 160, 0 }, // Layer 3
    { 0,   8,  16,  24,  32,  40,  48,  56,  64,  80,  96, 112, 128, 144, 160, 0 }, // Layer 2
    { 0,  32,  48,  56,  64,  80,  96, 112, 128, 144, 160, 176, 192, 224, 256, 0 }  // Layer 1
  },
  { // Version 1
    { 0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0, 0 }, // Reserved
    { 0,  32,  40,  48,  56,  64,  80,  96, 112, 128, 160, 192, 224, 256, 320, 0 }, // Layer 3
    { 0,  32,  48,  56,  64,  80,  96, 112, 128, 160, 192, 224, 256, 320, 384, 0 }, // Layer 2
    { 0,  32,  64,  96, 128, 160, 192, 224, 256, 288, 320, 352, 384, 416, 448, 0 }, // Layer 1
  }
};
```

If my understanding is correct MP3 is  MPEG-1 Audio Layer III or MPEG-2 Audio Layer III.

So, based on the table above mp3 should also support 192kHZ. But I admit my comment there is just bad! I will remove it.

Comment (plbossart) [reply src/audio/codec_adapter/codec_adapter.c:411]: you may be confusing kbs and kHz :-)

the rates are 
version 1: 48, 44.1 and 32kHz
version 2: 24, 22.05 and 16 kHz
version 2.5: 12, 11.025 and 8 kHz (this is actually not a standard but an extension by FhG).

https://en.wikipedia.org/wiki/MP3 has most of the needed information.

File: src/audio/codec_adapter/codec_adapter.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (dbaluta) [reply src/audio/codec_adapter/codec_adapter.c:413]: @plbossart Definitely! That was my next step. I try to do it gradually and also have something that's working for our releases. Customers want to look at the DSP even if it isn't the final code.
I will try to see how we satisfy both requirements. In time! :)

- the only thing that's *definitely* out of place right now is the m4 topology.

For the rest of the code:
- existing DAP support (based on PCM frames, should still work)
- no other functionality is impacted.

Next steps:
(1) decouple codec adapter from DAI scheduling. In fact here there is no clear line where should I do the decoupling.

Because codec_adapter has an output buffer that holds the decoded samples and DMA needs to copy based on DAI pace
each period "period_bytes" of data.
So, each DAI period I need to call codec_adapter_copy() to push data from codec_adapter local buffer to DMA buffer. 

Also, the Host need to push data to codec_adapter based on its own pace. So, Host should copy one "compress period" of data each time the DAI has rendered one mp3 frame.

So, in simple terms: 
- Host should copy one mp3 compressed frame each 0.026 sec (compress period will be 26ms).
- DAI should copy SAMPLE_RATE/1000 sampes each 1ms (compress period will be 1ms).

Also, we noticed that we could process around 3 MP3 frames with a single pipline, so there is enough room for our simple sceario to work.

File: src/audio/codec_adapter/codec_adapter.c (hunk 6 above)
Comment (dbaluta) [reply src/audio/codec_adapter/codec_adapter.c:411]: Yes, you are right the comment should be read 192kbps.

File: src/audio/codec_adapter/codec_adapter.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (plbossart) [reply src/audio/codec_adapter/codec_adapter.c:413]: I think you should really have 3 pipelines
a) dai with mixer
b) PCM for notifications/beeps that may occur at any time
c) compressed. That pipe may run at a lower priority or larger granularity than the DAI period. The last thing you want it to try and complete one decode at a 26x speed, you have to spread the work across multiple dai periods otherwise you *will*  underrun when your system is loaded.

IIRC in all implementations, the DAI typically works with O(ms) buffers, and the compressed pipeline can work with much larger frames. In fact if you look at newer codecs they will generate 4k or 8k samples in one frame. So you need some sort of ring buffer to adjust between codec and dai.

[the remark this answers is not in the document]
Comment (plbossart) [reply src/audio/codec_adapter/codec_adapter.c:413]: And I forget that you need an SRC on your compressed pipeline if you want to support all frequencies.

[the remark this answers is not in the document]
Comment (dbaluta) [reply src/audio/codec_adapter/codec_adapter.c:413]: Nice suggestion.

**Comment by lgirdwood** [comment]
Please use "main" as PR target.
